docs/operators/deduper.md

```diff
@@ -0,0 +1,523 @@
+-   ***timeExpression*** - This is the key which can help identify the expired records, as explained above. In this particular use case, it can be a timestamp field which indicates when the log message was generated.
+-   ***expireBefore*** - This is the period of expiry as explained above. In our particular use case this will be 24 hours.
+
+
+
+Configuration of these parameters would resolve this use
+case.
+
+Use cases - Summary
+================================
+
+1.  **Basic Dedup** - Deduplication of
+    bounded datasets. Data is assumed to be bounded. This use case is
+    not meant for never ending streams of data. For example:
+    Deduplication of master data like customer records, product catalogs
+    etc.
+2.  **Time Based Dedup** - Deduplication of
+    unlimited streams of data. This use case handles unbounded streams
+    of data and can run forever. An expiry key and criterion is expected
+    as part of the input which helps avoid storing all the unique data.
+    This helps speed up performance. Any Date field in the incoming
+    tuple can be used as a time based expiry key.
+    -  *With respect to system time* - Time progresses with system time. Any expiry criterions are executed with this notion of system time. This is possible if the incoming tuple does not have a time field, or the user does not specify this.
+    -  *With respect to tuple time* - Time progresses based on the time in the incoming tuples. Expiry criterions are executed with the notion of time indicated by the incoming tuple. The time field is mandatory for this scenario.
+
+Technical Architecture
+===================================
+Class Structure
+----------------------------
+
+![](images/deduper/image03.png)
+
+---------------
+
+Architectural Details
+----------------------------------
+
+![](images/deduper/image02.png)
+
+Concepts
+---------------------
+
+### Dedup Key - Specified by *keyExpression* parameter
+
+A dedup key is a set of one or more fields in the data tuple which
+acts as the key for the tuples.
+This is used by the deduper to compare tuples to arrive at the
+conclusion on whether two tuples are duplicates. If Dedup Key of two
+tuples match, then they are duplicates, else they are unique.
+
+### Expiry Key - Specified by *timeExpression* parameter
+
+A tuple may or may not have an Expiry Key. Dedup operator cannot
+keep storing all the data that is flowing into the operator. At some
+point it becomes essential to discard some of the historical tuples in
+interest of memory and efficiency.
+
+At the same time, tuples are expected to arrive at the Dedup
+operator within some time after they are generated. After this time, the
+tuples may be considered as stale or obsolete.
+
+In such cases, the Deduper chooses to consider these tuples
+expired and takes no action but to
```

Comment:
takes no action but to separate out => takes no action other than separating out